crates/core_simd/src/vector.rs

```diff
@@ -979,3 +1188,27 @@ where
 {
     type Mask = isize;
 }
+
+#[inline]
+fn lane_indices<T, const N: usize>() -> Simd<T, N>
+where
+    T: MaskElement + Default + core::convert::From<i8> + core::ops::Add<T, Output = T>,
+    LaneCount<N>: SupportedLaneCount,
+{
+    let mut index = [T::default(); N];
+    for i in 1..N {
+        index[i] = index[i - 1] + T::from(1);
+    }
+    Simd::from_array(index)
+}
+
+#[inline]
+fn mask_up_to<M, const N: usize>(len: usize) -> Mask<M, N>
+where
+    LaneCount<N>: SupportedLaneCount,
+    M: MaskElement,
+{
+    let index = lane_indices::<i8, N>();
```

Comment:
this should be `isize` instead of `i8`, since we will eventually have types with more than 127 elements.

Reply:
well, apparently LLVM is more terrible than I thought: https://rust.godbolt.org/z/vj75dsT51

Reply:
What about using M directly?  If type inference doesn't play nice, using intrinsics directly can probably get past it.

Reply:
Doing it on `M` would require `M: core::ops::TryFrom<usize>` which would appear on the public interface, and then there's still no way to convert `len: usize` to `M` safely, nor is there a `core::ops::Max` that would give you a generic `::MAX` constant. Am I missing anything?

Reply:
You can add hidden helper functions/consts like that (e.g. a `from_usize`) to the Sealed trait on MaskElement.  Not super elegant, but it will work and eliminate the bounds

Reply:
ok, I came up with some code that still works with arbitrarily many lanes and compiles to good code: https://rust.godbolt.org/z/Gqx85P4Ex

Reply:
the `M::Unsigned: SimdCast + Into<u64>` bounds should be added to `Sealed::Unsigned` so the user doesn't need to specify anything more than `M: MaskElement`

Reply:
@programmerjake I don't think I can add `Sealed::Unsigned: Into<u64>` because usize does not implement that trait.

```
error[E0277]: the trait bound `u64: From<usize>` is not satisfied
  --> crates\core_simd\src\masks.rs:96:24
   |
96 | impl_element! { isize, usize }
   |                        ^^^^^ the trait `From<usize>` is not implemented for `u64`, which is required by `usize: Into<u64>`
   |
```

Reply:
I don't think `where` bounds are picked up when you use the trait, they only constrain it.  I would just add functions to Sealed anyway, to avoid leaking bounds (that we can't remove in the future if we want to)

Reply:
I originally had that as `u128` instead of `u64`, though like caleb suggests, a Sealed method may be best.

Reply:
> I don't think `where` bounds are picked up when you use the trait, they only constrain it.

associated type bounds seem to be picked up: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4d88c89dd7ee65b133b92a3ccedca39d

Reply:
Pushed the change. I barely understand it but I think it should work?